src/gleam/time/duration.gleam

```diff
@@ -0,0 +1,80 @@
+import gleam/order
+
+// TODO: document
+pub opaque type Duration {
+  // When compiling to JavaScript ints have limited precision and size. This
+  // means that if we were to store the the timestamp in a single int the
+  // duration would not be able to represent very large or small durations.
+  // Durations are instead represented as a number of seconds and a number of
+  // nanoseconds.
+  //
+  // If you have manually adjusted the seconds and nanoseconds values the
+  // `normalise` function can be used to ensure the time is represented the
+  // intended way, with `nanoseconds` being positive and less than 1 second.
+  Duration(seconds: Int, nanoseconds: Int)
+}
+
+// TODO: test
+/// Ensure the duration is represented with `nanoseconds` being positive and
+/// less than 1 second.
+///
+/// This function does not change the amount of time that the duratoin refers
+/// to, it only adjusts the values used to represent the time.
+///
+fn normalise(duration: Duration) -> Duration {
+  todo
+}
+
+// TODO: docs
+// TODO: test
+pub fn compare(left: Duration, right: Duration) -> order.Order {
+  todo
+}
+
+// TODO: docs
+// TODO: test
+pub fn difference(left: Duration, right: Duration) -> Duration {
+  todo
+}
+
+// TODO: docs
+// TODO: test
+pub fn add(left: Duration, right: Duration) -> Duration {
+  todo
+}
+
+// TODO: docs
+// TODO: test
+pub fn to_iso8601_string(duration: Duration) -> String {
```

Comment:
Does it make sense for a duration to be formatted as a string like this? For example, if I calculate the difference between two timestamps and get 500 milliseconds as a duration, how can this be formatted in a semantically appropriate ISO 8601 string?

Reply:
ISO8601 is the name of the standard, not the timestamp format! That form defines multiple formats such as a timestamp format and a duration format.

https://en.wikipedia.org/wiki/ISO_8601#Durations

Reply:
Ah, did not realize ISO8601 had formatting for durations, cool!